Sage: Fix on Darwin

Feyorsh opened the pull request:
## Description of changes

There shouldn't be any fundamental blockers to building Sage on aarch64-darwin, given that it builds fine with Homebrew. *I am quite inexperienced with Nix, so please let me know if I need to fix my code to be more idiomatic or otherwise correct.*
Currently, I am able to build Sage and "pass" all the test cases: some test cases fail because of linker warnings when Cython compiles:
```
sage -t --long --random-seed=188234609292076795741796414804145521592 /nix/store/b79f4fps121z8lfdy3km6x51bb2xgiqc-sage-src-10.0/src/sage/symbolic/pynac.pxi
**********************************************************************
File "/nix/store/b79f4fps121z8lfdy3km6x51bb2xgiqc-sage-src-10.0/src/sage/symbolic/pynac.pxi", line 6, in sage.symbolic.pynac
Failed example:
    cython(  # optional - sage.misc.cython
    '''
    cimport sage.symbolic.expression
    ''')
Expected nothing
Got:
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-zlib-1.3/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-bzip2-1.0.8/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-expat-2.5.0/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-xz-5.4.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-libffi-3.4.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-gdbm-1.23/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-sqlite-3.43.1/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-readline-8.2p1/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-ncurses-6.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-openssl-3.0.11/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-apple-framework-SystemConfiguration-11.0.0/lib'
    ld: warning: directory not found for option '-L/nix/store/8ldla7hgqh9nfdmcxb7h41xkh5lrgfbl-tzdata-2023c/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-zlib-1.3/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-bzip2-1.0.8/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-expat-2.5.0/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-xz-5.4.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-libffi-3.4.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-gdbm-1.23/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-sqlite-3.43.1/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-readline-8.2p1/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-ncurses-6.4/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-openssl-3.0.11/lib'
    ld: warning: directory not found for option '-L/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-apple-framework-SystemConfiguration-11.0.0/lib'
    ld: warning: directory not found for option '-L/nix/store/8ldla7hgqh9nfdmcxb7h41xkh5lrgfbl-tzdata-2023c/lib'
    ld: warning: directory not found for option '-L/nix/store/2bbhxavs7l15k3rh72rlg2cp2iwlibvp-jmol-16.1.41/lib'
**********************************************************************
1 item had failures:
   1 of   2 in sage.symbolic.pynac
    [1 test, 1 failure, 18.40 s]
```
I'm assuming I need to poke around sage-env.nix as suggested [here](https://github.com/NixOS/nixpkgs/blob/b3e0dee4ee7ac77c712b372158256e62318e6651/pkgs/applications/science/math/sage/sage-env.nix#L107) and add these missing dependencies to `PKG_CONFIG_PATH`. Besides that, **Sage behaves as you would expect, all functional test cases pass just fine.** Disabling tests is obviously not an actual solution, but I'm confident these issues should be somewhat straightforward to fix.

108e281: fflas-ffpack seems to build and pass all test cases successfully.
d154d8a32f031930d8f2ba80f9f60866fdf0d142: Interestingly, although m4ri builds fine, m4rie seems to only build properly on Apple Clang, not LLVM Clang. As suggested in an upstream [issue](https://bitbucket.org/malb/m4rie/issues/23/trying-to-compile-on-apple-m1), I changed `-O2` to `-O0` and it compiles. Please let me know if I should be modifying CFLAGS in a different way or changing hardening flags.
98a725a: There seems to be an issue with Apple's `libintl` assuming locale variables being set, which causes most of Giac's test cases to fail. Passing `--disable-nls` partially fixes the issue, leading to only 4 test cases failing. I would like to fix this completely, but according to a Sage [discussion](https://github.com/sagemath/sage/issues/34537#issuecomment-1418187176), it looks like these failures are not new. I also added some Darwin patches for Giac from Sage, but they likely have no bearing on whether or not the above issue gets resolved.
aff699e: When running Sage, I received warnings about `awk` not being found. Indeed, `gawk` was not an input into `sage-env.nix`. It seems surprising that only Darwin would require `awk` so I didn't guard it with `lib.optionals stdenv.isDarwin`; is this warning present on Linux without the inclusion of `gawk`?
c65031c8be5e03f4a9258c176da990e42e99ddca: I just used `macosx` as the target because it's what Sage uses and it looks like there's some issues with the multithreaded build on aarch64-linux. I haven't tested `macosx-thr`, though.
7e419ac: fpylll (and fplll) need to disable `long double` on both Cygwin __and__ aarch64-darwin (x86_64-darwin does support `long double` and shouldn't be changed).
### TODO
**linbox**: I noticed that a test failed when building linbox as part of Sage (i.e. `nix-build -A sage`), but when I built it separately all tests passed. I assumed that was just an issue with a test case running out of memory or timing out, but I don't know how to confirm that that's the case.
**sympow**: Install check fails, almost certainly due to this warning:
```
CFLAGS for SYMPOW:  -std=gnu17 -fno-fast-math -ffp-contract=on  -O0
The double precision of your FPU is 105 bits.
Error: the Quad Double library used by SYMPOW assumes IEEE-754 double
precision numbers with exactly 53 bits in the mantissa (64 bits in
total).  Unfortunately, this is not the case on your system and we
currently have no workaround for your system.  Running SYMPOW will
almost certainly fail on some inputs.
```
Sage [packaging](https://github.com/sagemath/sage/blob/07a2afd65fb4b0a1c9cbc43ede7d4a18c921a000/build/pkgs/sympow/spkg-install.in) does not appear to fix this, so the "fix" might be simply disabling Sympow within Sage.
**Giac**: Pass remaining test cases (or ignore if infeasible).
**Sage**: Fix linker warnings as described above.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [X] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Feyorsh committed:
giac: 1.9.0-43 -> 1.9.0-993

Nearly fixed on aarch64-darwin! (Note that this coincides with a
release of Xcas for macOS, but this is not using Parisse's tarball).

collares commented:
Thanks, this is much appreciated! 

Just a heads-up: I disabled Singular docbuilding on darwin last year because it was hanging and I didn't have a machine to debug it on. I would expect this to cause a Sage test failure or two, but don't worry about it if it doesn't cause test failures.

symphorien reviewed (commented):
xcas works fine

Feyorsh commented on pkgs/development/libraries/science/math/m4rie/default.nix:
```diff
@@ -21,6 +21,7 @@ stdenv.mkDerivation rec {
     m4ri
   ];
 
+  makeFlags = [] ++ lib.optionals stdenv.isDarwin [ "CFLAGS=-O0" ];
```
Gotcha. Could you please try to build m4rie using `clangStdenv` to see if this codegen issue affects Linux too? This one-liner should do the trick: `nix-build -E "with import <nixpkgs> {}; pkgs.m4rie.override{ stdenv = pkgs.clangStdenv; }"`

Ex-32 replied:
Just ran this oneliner on my x86_64-linux machine, worked fine, all the tests passed.

Feyorsh commented:
Fixed **fpylll** by disabling `HAVE_LONG_DOUBLE` in both fplll and fpylll. I backported a patch from the 0.6.0 release of fpylll because updating would require switching to Cython3, which causes conflicts with other packages in Sage that use Cython2. We should update all packages to Cython3 at the same time when Sage adds support for it, probably as part of the [10.2](https://github.com/sagemath/sage/wiki/Sage-10.2-Release-Tour) release.

collares commented:
Agreed on the Cython3 point! By the way, feel free to ping me when you remove the draft status from the PR (since GitHub does not send notifications about it).

7c6f434c commented:
@Feyorsh Speaking of 10.2, @collares has just performed and merged this update.

Feyorsh commented:
@7c6f434c Thanks for the heads up — apologies for letting this sit in stasis. The only thing I couldn't fix were the Cython linker warnings (and sympow, which fundamentally cannot work on aarch64-darwin), but I will try to rebase on top of the 10.2 update to fix that.

Feyorsh commented:
Ok, I finally came back to this. There are several more changes that have to be made due to `stdenv` updating to Clang 16, which promotes a few warnings into errors and compiles with C++17 by default (at least I'm pretty sure it's the default...)

I initially tried compiling with C++14, but that's frankly a band-aid fix; the only packages I still need to fix are Giac and Sage itself (https://github.com/sagemath/sage/issues/36840).

A few notes:
- d364c83: What's the policy for deletion of packages? As far as I can tell Sage is the only user of FlintQS, but presumably we need to deprecate it instead of deleting it outright.
- 8d1b4242327ce5be15e9207aab4908b8eddb1e1f: This is somewhat inelegant, but I see no better way to get the job done (I suppose I could just upstream the line ending change...)
- cfaf402: I'm going to revert this and bump Giac to 1.9.0-93 and fix any other errors.
- 08988c0: For future reference: Is there any issue fetching patches from PRs? Unsure if these have the same permalink guarantees as commits
- 6b05225: The `src` tarball is *not* the same as the Git repo, so I had to modify a patch slightly.

Also: I still don't know how to fix the `ld: warning: directory not found for option '-L/nix/store/...` warnings. Any idea where linker flags are set besides `buildInputs`?

symphorien commented:
> https://github.com/NixOS/nixpkgs/commit/08988c0621692344bb489e8b80ee2212fd0b2377: For future reference: Is there any issue fetching patches from PRs? Unsure if these have the same permalink guarantees as commits

no they don't. if the PR is force pushed, then the hash is invalidated. You can fetch the commit of the PR, but even though the commit hash cannot be invalidated, the commit can be gc-ed and disappear if the PR is closed or force-pushed. So for non-merged-yet PRs the only truly future proof way is vendoring the patch.

symphorien commented:
> https://github.com/NixOS/nixpkgs/commit/d364c8370393ff7923c4510eb7df102c28ff084f: What's the policy for deletion of packages? As far as I can tell Sage is the only user of FlintQS, but presumably we need to deprecate it instead of deleting it outright.

you are supposed to add a line in pkgs/top-level/aliases.nix I think.

collares commented:
>     * [6b05225](https://github.com/NixOS/nixpkgs/commit/6b05225e362d7e004f854949742944101fdf3821): The `src` tarball is _not_ the same as the Git repo, so I had to modify a patch slightly.

If the only modification was deleting the `.gitignore` changes, you can pass `excludes = [".gitignore"]` to `fetchpatch`. 

Thanks for the update, by the way! The changes here are looking good. I will probably review them in more detail on the weekend. I also just merged the Sage 10.3 upgrade, which included a Singular bump to 4.3.2p16.

Feyorsh commented:
I did not, thanks for the heads up.

Feyorsh committed:
sage: fixes for LLVM stdenv

Also backport a fix that silences `xcode-select: not found` warnings

Feyorsh commented:
Having some issues building Singular *only* when building Sage (something about an invalid `ioctl`: it might be related to the PS generation/`xdg-open`?).
It appears the only difference is `flint3` is overridden without BLAS, but I'm getting docbuild errors that seem to be coming from GFAN... not only that, but GFAN appears to be the primary source of doctest errors in Sage (might be related to building with C++0X).

Also, Giac is still borked, and I've still yet to solve the mystery of the extra linker flags.

collares commented:
Some commits of this PR seem ready, even if the full objective of building Sage is not yet complete. Would you mind splitting mergeable commits to a separate PR? This way, I can review and merge them sooner. 

By the way, I don't have a Darwin machine, so all I see on the Hydra side is a flint3 build failure. Being able to test things on ofBorg would be helpful for me. ioctl errors while building the Singular documentation are a quirk of how the docs get built. Once you know which doc is failing to build, you can manually run the same Singular invocation to take a proper look at the error. `breakpointHook` can be helpful to look at failed build states, it is described in the Nix manual.

(By the way, there's a comment on the Singular package about using texinfo4, and it no longer makes sense after the move to texinfo. Can you delete the comment? Thanks!)

collares commented:
Maybe https://github.com/sagemath/sage/blob/develop/build/pkgs/gfan/patches/nodel.patch helps the gfan issue

Feyorsh committed:
gap: fix build on darwin

Feyorsh committed:
gfan: fix to pass sage doctests on clang

This issue does not present at build time, but only in sage
doctests. See https://www.github.com/sagemath/sage/issues/25118

Feyorsh committed:
linbox: drop `withSage`

--enable-sage was removed a while ago
https://github.com/linbox-team/linbox/pull/146

Feyorsh commented:
Ok, hopefully this is the last of it... just in time for Sage 10.4. Only 26 doctest failures.
## Linker warnings
Basically, the Python derivation litters a bunch of dead links in its `sysconfig` because it uses `nukeRefs`:
```
nix-shell -p python3 --command "python3 -m sysconfig"
```
Now, this isn't an issue when using GNU binutils because the `directory not found` warnings aren't emitted (by default, at least), but that's not the case on Darwin:
```
# cat $(nix-store --realize $(nix-instantiate -E --argstr system aarch64-darwin "{system ? builtins.defaultSystem }: with import <nixpkgs> {inherit system;}; runCommandCC ''test'' {} ''mkdir -p \$out; echo \"int main(){}\" | cc -x c - -Lnil &> \$out/ld_test''"))/ld_test
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
ld: warning: directory not found for option '-Lnil'

# cat $(nix-store --realize $(nix-instantiate -E --argstr system aarch64-linux "{system ? builtins.defaultSystem }: with import <nixpkgs> {inherit system;}; runCommandCC ''test'' {} ''mkdir -p \$out; echo \"int main(){}\" | cc -x c - -Lnil &> \$out/ld_test''"))/ld_test
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
```

So, how do we fix this? Monkeying around with the `ld-wrapper` is probably a bad idea; trying to properly cleanup python's sysconfig might be good for nixpkgs as a whole, but that's far beyond the scope of this PR. Although I wasn't able to apply a filter to on the specifically problematic compiler output, I _was_ able to pass `-w` to the linker to get it to shut up. Now, this is obviously not ideal as it might make it harder to spot a linker problem, but given that this is constrained just to Sage's Cython, it might be good enough for now.

## Packages
**Sage** Backported a few patches from 10.4, which I guess aren't going to stay backported for long. Fixed the linker warning problem; only doctests currently failing are Singular.

**Giac** Updated to 1.9.0-993; made it compile with `-std=c++17` and fixed a bug affecting all platforms when building without FLTK (I emailed both patches upstream to Parisse, in case he's interested). The tests requiring the GUI still fail and there are two other failures on `aarch64-darwin`, likely due to the jank way multiple Giac instances share the same PARI instance (in `pari.cc.old`; other platforms use `pari.cc` which doesn't have this issue). This is a big improvement from 1.9.0-43, although I did hit this minor bug that causes a doctest failure:
```
1>> sum(1/(1+k^2), k, -infinity, +infinity)
Warning, infinity is unsigned, perhaps you meant +infinity
Warning, infinity is unsigned, perhaps you meant +infinity
Warning, infinity is unsigned, perhaps you meant +infinity
0

File "/nix/store/0vwkw5ks7lzdi2rqvf2zfc6pzgqg4a61-sage-src-10.3/src/sage/symbolic/expression.pyx", line 13094, in sage.symbolic.expression.Expression.sum
Failed example:
    (sum(1/(1+k^2), k, -oo, oo, algorithm = 'giac')).factor()
Expected:
    pi*(e^(2*pi) + 1)/((e^pi + 1)*(e^pi - 1))
Got:
    0
```

**Linbox** There's the aforementioned "heisenbug" of a failing doctest, but I can't reproduce it. I turned on core dumping recently so maybe I'll get lucky and be able to get a stack trace ;)

**GAP** Dropped the upgrade to 13, but good news is I already reported the only error preventing that from building upstream and its been fixed. Less work for the future.

**Singular** So the docbuilding is failing because of these pesky `system` calls invoking a LaTeX compiler and invoking `xdg-open` on the result. I actually don't  know why this works on any platform, given `texdraw.sty` isn't even in `texliveSmall`. I tried adding `xdg-utils` and substituting store paths in --- no dice. I tried removing the code path that calls LaTeX altogether --- still nothing. I can confirm the hang when running the doctest manually, and it looks like any `system` call in Singular is faulty... I need to investigate further, and I'll try upgrading to 4.4.0 to see if that makes a difference.

**Sympow** Still haven't fixed it; I added in the checks from Sage's `spkg-check.in`, but I frankly have no idea how to fix the error (or even how to make sense of what's going wrong). Sage doctests pass, so I'm inclined to call this "good enough" for hopefully marking sage as unbroken, but it would be nice to figure out if/how this can be put to rest. (There's also a newer version, doesn't fix the problem though).

(I'm also happy to spin some of these off into a 10.4 upgrade PR)

Feyorsh committed:
singular: fix docbuilding on non-sandboxed builds

This upgrade fixes the issue with drawTropicalCurve docbuilding
on Darwin (although said doctest is pretty useless now that Apple has
removed postscript from macOS).

Removed enableGfanlib; you can still overrideAttrs

Added mpfr to be more consistent with upstream packaging
recommendations

Feyorsh committed:
sympow: add sage self tests

Feyorsh commented:
Singular docbuilding fails on Darwin because of issues with `/tmp`. As part of building the docs, Singular creates and _does not_ cleanup a specific path in `/tmp`, which causes a collision when the next build tries to write to the same file (Linux doesn't have this issue because better sandboxing). This really ought to be fixed [upstream](https://github.com/Singular/Singular/issues/1137), but for the time being I'll clean up the paths myself in the installPhase.

Turns out the Giac bug has already been fixed in the [Geogebra dev repo](https://github.com/geogebra/giac) (should we switch over to that at some point?).

Sage's Singular doctests are currently failing on `aarch64-darwin`, I'm looking into it. This is the only thing left to address on Darwin, all tests pass when I tested on `x86_64-linux`.

Feyorsh commented:
Ok, Singular doctest failure is probably a bug in Sage:
```
./result/bin/sage -t --long --randorder=111111 /nix/store/v6r22ncwk52vrll9ymia1sjgcff3kcbq-sage-src-10.3/src/sage/interfaces/singular.py
```
The failures themselves are indicative of being run out of order; I don't know why I'm able to hit this so consistently only when building on my `aarch64-darwin` box; I've never hit this when building from `x86_64-linux`, or when running doctests from sage on either host. I am able to reproduce the error anywhere with `--randorder` though.

I'd like some guidance on what to do here... disable the doctest because it works fine in interactive use? Wait for an upstream patch (this is probably quite involved)? Even though this is not quite resolved, I think it might be beneficial to merge this in as marked broken.

Feyorsh commented:
> merge this in as marked broken.

Sorry, I meant keep `meta.platforms = lib.platforms.linux`.

Feyorsh commented:
@collares @symphorien (there are nontrivial xcas changes, see above). Sorry this took so long.

Feyorsh commented:
Aha. So the Singular Sage failures are a result of Singular running out of stack; initially I saw this backtrace and thought it was a bug in Singular:
```
(lldb) bt
* thread #1
  * frame #0: 0x0000000189deea60 libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x0000000189e26c20 libsystem_pthread.dylib`pthread_kill + 288
    frame #2: 0x0000000189d33ab4 libsystem_c.dylib`__abort + 136
    frame #3: 0x0000000189d24694 libsystem_c.dylib`__stack_chk_fail + 96
    frame #4: 0x0000000102f4d6f4 libsingular_resources-4.3.2.dylib`feResource(feResourceConfig_s*, int) + 900
```
But then I tried disabling just the `singular_version` doctests, and there were no problems. Well, it's quite unlikely there's a bug in something as simple as printing a version string... but that's one big string (7562 bytes on my machine)! Normally that has nothing to do with the stack, but because Singular uses their own allocator I'm not so sure. Long story short increasing the stack size causes the test to consistently pass without fail; it seems `ulimit` is sort of frowned upon in nixpkgs because it's not reproducible, but let me know if you have a better idea for this test.

Feyorsh committed:
sage: fix singular doctests

Tampering with ulimit is not ideal and may not work in the Darwin
sandbox, but is necessary due to how Singular handles allocations.

Feyorsh committed:
sage: add aarch64-darwin to supported platforms

Feyorsh commented:
`ulimit` didn't work consistently, so I just disabled the version doctests.

symphorien commented:
You wrote large (but simple) patches to fix the build of giac with C++17. I think it would be both simpler and more maintainable to tell the compiler to use an older c++ standard instead, and minimize patching. See https://github.com/NixOS/nixpkgs/pull/309356 for an example.

Feyorsh commented:
The reason I feel more comfortable with this approach is because I 
am currently working with the Giac maintainer on upstreaming both 
patches; these can safely be dropped whenever Giac gets updated 
next, so I don't think there's a maintenance burden.

On Mon, Jul 22, 2024 at 13:39:22 EST Guillaume Girol wrote:

> You wrote large (but simple) patches to fix the build of giac 
> with
> C++17. I think it would be both simpler and more maintainable to 
> tell
> the compiler to use an older c++ standard instead, and minimize
> patching. See https://github.com/NixOS/nixpkgs/pull/309356 for 
> an
> example.

symphorien commented:
In that case the changes are fine. Thank you for your efforts!

collares commented:
I will leave a more detailed review in the next few days, but so far I just wanted to thank you for the thought and effort put into supporting Darwin properly!

Two minor things: If possible, I'd like the gfan patch and the disabling of Singular doctests to apply only on Darwin, because they have (very minor) downsides. Also, I'd consider removing the `/tmp` cleanup from Singular's installPhase once you're done testing your changes, since I think Hydra uses a sandbox on all platforms.

StefanoDeVuono commented:
@Feyorsh It looks like the tmp issue with Singular has been stalled a bit. If you're down, maybe we could collab on a PR to submit to them?

I'm looking forward to Sage on Darwin through Nix!

Feyorsh commented:
@StefanoDeVuono Admittedly I'm lacking the motivation to see such a PR through — even if the bug with `/tmp` was fixed, the LaTeX interface in [`tropical.lib`](https://github.com/Singular/Singular/blob/a1ac996d26e5c02ba18ad54e412b5ff32d1eb162/Singular/LIB/tropical.lib#L1857) is probably overdue for a rewrite anyway. I initially patched in the xdg-open shim for macOS, then I realized macOS dropped Postscript support a year ago. I could bundle in Ghostscript, but it seemed like an increasingly convoluted solution to support a very niche feature.

I don't mean to discourage you from fixing it upstream though... maybe reach out to the Gentoo Sage maintainer who opened the issue?

(I'm also going to be AFK for the next two weeks, if this PR gets reviewed in that time. Would love to see this land before the upgrade to 10.4!)

collares commented on pkgs/applications/science/math/giac/default.nix:
```diff
@@ -9,18 +9,28 @@ assert (!blas.isILP64) && (!lapack.isILP64);
   };
 
   patches = [
+    ./remove-old-functional-patterns.patch
```
Seems like your patch was upstreamed in the meantime! Thank you.

https://github.com/geogebra/giac/commit/e1f2e8899f7a579168f5565a05adf987457ddcff
https://github.com/geogebra/giac/commit/2f6ea6cd85ccc224d5ce2f8947ecbd2cc7840f1b

collares commented on pkgs/development/libraries/science/math/sympow/default.nix:
```diff
@@ -61,6 +61,11 @@ stdenv.mkDerivation rec {
   doInstallCheck = true;
   installCheckPhase = ''
     export HOME=$TMPDIR
+    "$out/bin/sympow" -curve "[1,2,3,4,5]" -moddeg | grep 'Modular Degree is 464'
+    echo "[1,-1,0,-79,289]" | "$out/bin/sympow" -analrank | grep ^"Analytic Rank is 4"
+    "$out/bin/sympow" -curve "[1,-1,0,-79,289]" -analrank | grep ^"Analytic Rank is 4"
+    "$out/bin/sympow" -curve "[0,1,1,-2,0]" -analrank | grep ^"Analytic Rank is 2"
+  '' + lib.optionalString (!stdenv.isAarch64) ''
     "$out/bin/sympow" -sp 2p16 -curve "[1,2,3,4,5]" | grep '8.3705'
   '';
```
```suggestion
  installCheckPhase =
    ''
      export HOME=$TMPDIR
      "$out/bin/sympow" -curve "[1,2,3,4,5]" -moddeg | grep 'Modular Degree is 464'
      echo "[1,-1,0,-79,289]" | "$out/bin/sympow" -analrank | grep ^"Analytic Rank is 4"
      "$out/bin/sympow" -curve "[1,-1,0,-79,289]" -analrank | grep ^"Analytic Rank is 4"
      "$out/bin/sympow" -curve "[0,1,1,-2,0]" -analrank | grep ^"Analytic Rank is 2"
    ''
    + lib.optionalString (!stdenv.isAarch64) ''
      "$out/bin/sympow" -sp 2p16 -curve "[1,2,3,4,5]" | grep '8.3705'
    '';
```
to appease our nixfmt overlords.

collares reviewed (commented):
Many apologies for taking so long to review your changes, the last few weeks were quite hectic on my side. Everything looks great! I only found a few minor nits. Happy to merge this PR when they're addressed.

collares commented:
@ofborg build sageWithDoc giac giac-with-xcas

Feyorsh commented on pkgs/applications/science/math/giac/default.nix:
```diff
@@ -129,7 +148,8 @@ stdenv.mkDerivation rec {
     platforms = platforms.linux ++ (optionals (!enableGUI) platforms.darwin);
-    broken = stdenv.isDarwin && stdenv.isAarch64;
+    # still failing 2 testcases; see https://github.com/NixOS/nixpkgs/pull/264126
+    # broken = stdenv.isDarwin && stdenv.isAarch64;
```
Hopefully with the push upstream to officially support macOS, the jank codepaths in `pari.cc.old` will be dropped and this is no longer an issue.

collares commented:
I think the `sympow` suggestion wasn't applied verbatim, because there's still a red X on this PR. I don't mind alternate formatting as long as we get a green tick in the `nixfmt-check` PR check. You can see the GitHub CI log for instructions on how to run `nixfmt-rfc-check`.

More worryingly are the Giac-related sage test failures. For example:

```
**********************************************************************
File "/nix/store/qb6ad55k9xdrpy44hx3by2gxk7r3grqz-sage-src-10.3/src/sage/libs/giac/giac.pyx", line 303, in sage.libs.giac.giac._giac
Failed example:
    A[2,3] = 33; A[0,2] = '2/7' # set non zero entries of the sparse matrix
Exception raised:
    Traceback (most recent call last):
      File "/nix/store/yihsym29ypwq1l41ys0phw5wa0hl136i-python3-3.12.4-env/lib/python3.12/site-packages/sage/doctest/forker.py", line 712, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/nix/store/yihsym29ypwq1l41ys0phw5wa0hl136i-python3-3.12.4-env/lib/python3.12/site-packages/sage/doctest/forker.py", line 1147, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.libs.giac.giac._giac[29]>", line 1, in <module>
        A[Integer(2),Integer(3)] = Integer(33); A[Integer(0),Integer(2)] = '2/7' # set non zero entries of the sparse matrix
                                                ~^^^^^^^^^^^^^^^^^^^^^^^
      File "sage/libs/giac/giac.pyx", line 1042, in sage.libs.giac.giac.Pygen.__setitem__ (build/cythonized/sage/libs/giac/giac.cpp:164286)
        GIAC_sto(v,g,1,context_ptr)
    RuntimeError: bad_function_call
```
and
```
sage -t --long --random-seed=220277908603107882552103743915235334071 /nix/store/qb6ad55k9xdrpy44hx3by2gxk7r3grqz-sage-src-10.3/src/sage/calculus/calculus.py
**********************************************************************
File "/nix/store/qb6ad55k9xdrpy44hx3by2gxk7r3grqz-sage-src-10.3/src/sage/calculus/calculus.py", line 1855, in sage.calculus.calculus.inverse_laplace
Failed example:
    inverse_laplace((2*s^2*exp(-2*s) - exp(-s))/(s^3+1), s, t,
                    algorithm='giac')
Expected:
    -1/3*(sqrt(3)*e^(1/2*t - 1/2)*sin(1/2*sqrt(3)*(t - 1))
     - cos(1/2*sqrt(3)*(t - 1))*e^(1/2*t - 1/2) + e^(-t + 1))*heaviside(t - 1)
     + 2/3*(2*cos(1/2*sqrt(3)*(t - 2))*e^(1/2*t - 1) + e^(-t + 2))*heaviside(t - 2)
Got:
    Giac crashed -- automatically restarting.
    sage34*dirac_delta(sage36)
```
It's possible one of the errors is a side-effect of the other, but at least one should be real.

In unrelated news, unfortunately Sage recently broke on master due to the Cython 3.0.11 upgrade and I might have to update it to 10.4 to apply https://github.com/sagemath/sage/pull/38500. But if Sage tests pass on this branch, I don't see any problem in merging this PR. Note that you'll have to autosquash the fixup commits, since GitHub does not do that automatically.

collares commented:
@Feyorsh Turns out the upstreamed Giac patches I mentioned (which you migrated to in https://github.com/NixOS/nixpkgs/pull/264126/commits/1d1710cd34a47eb75db7bacc812a8e927f1e3cf6) do not exactly match your previous version, and everything works before the change I suggested. Apologies, I should have tested it before making the suggestion.

If you reinstate `remove-old-functional-patterns.patch`, fix the sympow formatting error (as determined by the automatic checks) and autosquash the fixups, I can merge this. Let me know if you prefer that I do it instead (I'd be happy to).

Feyorsh commented:
That's on me for not running the Sage tests before pushing. Based on my correspondence with Parisse, it seems my patches were more of a loose inspiration than anything else (let's just hope it works when Giac gets bumped in the future). I'm running a full build currently and I'll let you know if I encounter any issues with the Sage test suite; if not, this should be good to go.

Feyorsh commented:
> I'm running a full build currently and I'll let you know if I encounter any issues

lgtm @collares

collares commented:
@ofborg build giac-with-xcas sageWithDoc

collares commented:
Wonderful, thank you very much! Excited to have Darwin support for Nixpkgs Sage.

collares commented:
@Feyorsh Unfortunately Maxima failed to build on aarch64-darwin: https://cache.nixos.org/log/jq3wzld79yzdqaxqlvlsyfa1x7nm8vk0-maxima-5.47.0.drv

Feyorsh commented:
Looking into it. Can't reproduce the error with `maxima` in the sandbox, so I suspect one of the Sage-specific patches was made obsolete with the bump to `5.47`.

collares commented:
Are you testing Maxima as it is built for Sage? That is,

 `pkgs.maxima-ecl.override { lisp-compiler = pkgs.ecl.override { threadSupport = false; useBoehmgc = true; }; }`

In particular, ECL as the lisp compiler seems to always use GCC as part of the Lisp compilation process (as seen below):

```
  postInstall = ''
    sed -e 's/@[-a-zA-Z_]*@//g' -i $out/bin/ecl-config
    wrapProgram "$out/bin/ecl" --prefix PATH ':' "${
      lib.makeBinPath [
        gcc                   # for the C compiler
        gcc.bintools.bintools # for ar
      ]
    }"
  '';
```

If that's not the issue and you're not seeing the error even in this configuratoin, my guess would be it is an intermittent GC problem (maybe related to https://gitlab.com/embeddable-common-lisp/ecl/-/issues/718 or https://github.com/ivmai/bdwgc/issues/569#issuecomment-1732545197, although in our case it wouldn't be happening at ECL build time.).

Feyorsh commented:
Yes, I tested using 74b34d69ee14 per the Hydra reproduce script, with `maxima.override { lisp-compiler = ecl.override { threadSupport = false; useBoehmgc = true; }; }` (expression is slightly different as I built the top level derivation instead of through Sage). It might be difficult to reproduce this locally if the Heisenbug only appears under load.

Unfortunately I need another 5 days until I can give this proper attention; this issue doesn't _seem_ to be a regression caused by this PR, but if these commits need to be reverted, so be it. Additional support should not come at the expense of breaking existing packages.

collares commented:
Oh, sorry for giving the wrong impression! Fixing the Maxima issue is not urgent, and this PR does not need to be reverted. `maxima-ecl` itself didn't break, the failure was just for the custom version of Maxima used on Sage, and just on aarch64-darwin.

collares commented:
Also not urgent, just for your information: Sage Maxima built OK on staging-next, so it seems like it was just an intermittent failure. But Singular is failing with

```
**** Error: while running example drawNewtonSubdivision from d2t_singular/tropical_lib.doc:665.
Call: '../Singular/Singular  -teqr12345678 --no-rc ./examples/drawNewtonSubdivision.sing > ./examples/drawNewtonSubdivision.res'
make[4]: *** [Makefile-docbuild:180: d2t_singular/tropical_lib.tex] Error 1

**** Error: non-zero exit status of system call: 'make -f Makefile-docbuild  --no-print-directory -s TAG='-tag gfan' VERBOSE=1 ./d2t_singular/tropical_lib.tex': Inappropriate ioctl for device
```

This seems less likely to be an intermittent failure, but perhaps we should wait for a rebuild. See https://hydra.nixos.org/build/271681864 for details.
